src/common/utility.cpp

```diff
@@ -0,0 +1,72 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// utility.cpp
+//
+// Identification: src/include/common/utility.h
+//
+// Copyright (c) 2015-17, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include <unistd.h>
+#include <vector>
+#include <string.h>
+#include <errno.h>
+
+#include "common/utility.h"
+#if __APPLE__
+extern "C"{
+#include <sys/cdefs.h>
+int close$NOCANCEL(int);
+};
+#endif
+
+namespace peloton {
+
+
+int peloton_close(int fd, int failure_log_level) {
+  int close_ret = -1;
+#if __APPLE__
+  close_ret = ::close$NOCANCEL(fd);
+#else
+  close_ret = close(fd);
+#endif
+
+  if (close_ret != 0) {
+    std::vector<char> buffer(100, '\0');
+    int saved_errno = errno;
+    char *error_message = nullptr;
+#if __APPLE__
+    (void)strerror_r(errno, buffer.data(), buffer.size() - 1);
+    error_message = buffer.data();
+#else
+    error_message = strerror_r(saved_errno, buffer.data(), buffer.size() - 1);
+#endif
+    (void) error_message;
+
+    const char *format_string = "Close failed on fd: %d, errno: %d [%s]";
+
+    switch (failure_log_level) {
```

Comment:
What is this `switch` needed for? Is there actually the need for calling the function and having the output on different log levels?
If I understand the problem correctly, #1269 is about whether we expect the connection to  be closed already or not. In that case I would suggest to put exactly this semantic in the function, and only write the message, if something happened that we didn't expect. (e.g. a flag `bool expect_still_open=true`)

Reply:
1.switch-case has been removed.
2.I think the intention is to call close on a fd once and only once. On return of 'close', the fd is guaranteed to be closed anyway by OS even if return value is -1. Leave a hint message on failure for debug purpose.